### PR TITLE
Added Spectral Norm

### PR DESCRIPTION
**Pull request by lavoiems**
Added SpectralNormLayer. SpectralNormLayer is a normalization technique defined here: https://openreview.net/forum?id=B1QRgziT-

The strategy used to implement a wrapper over some pytorch layer. I added an option to use the spectralnorm on SimpleConvnet and added an option to use it gan.

### CODE REVIEW

**Comment by rdevon** [comment]
@IshmaelBelghazi Can you take a quick look at this?

**Comment by rdevon** [comment]
There's conflicts how due to updates to the framework. Fixing this should be easy. Once this is done, I will do the review.

File: models/modules/SpectralNormLayer.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    sigma = u.dot(weight.data.view(height, -1).mv(v))
+    return torch.div(weight, sigma), u
+
+
+class SNConv2d(nn.Conv2d):
+    def __init__(self, *args, n_power_iterations=1, **kwargs):
+        super(SNConv2d, self).__init__(*args, **kwargs)
+        self.n_power_iterations = n_power_iterations
+        self.height = self.weight.data.shape[0]
```
Comment (remilepriol) [inline models/modules/SpectralNormLayer.py]: Explain in a comme t that the authors of SN choose to regularize convolutional layers 
representing them as a matrix with height = batch size.

Comment (lavoiems) [reply models/modules/SpectralNormLayer.py]: If I understand correctly your question, you should find the answer to your question here: https://openreview.net/forum?id=B1QRgziT-&noteId=SJmRwz7xG

Comment (remilepriol) [reply models/modules/SpectralNormLayer.py]: I know. I just think it would be good to explain how this is a choice in a comment. The true method would be the one suggested by Goodfellow.

Comment (lavoiems) [reply models/modules/SpectralNormLayer.py]: Good point. I will add your suggestion and update this PR to the flavor of master.

**Review by remilepriol — COMMENTED** [review]
Faithful port of the author's implementation to pytorch.

**Review by remilepriol — APPROVED** [review]
